src/NHibernate.Test/NHSpecificTest/GH2750/FixtureByCode.cs

```diff
@@ -0,0 +1,179 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Transactions;
+using NHibernate.Cfg;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH2750
+{
+	[TestFixture]
+	public class ByCodeFixture : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<TestEntity>(rc => { rc.Id(x => x.Id, m => m.Generator(Generators.Assigned)); });
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override void Configure(Configuration configuration)
+		{
+			configuration.SetProperty(Cfg.Environment.BatchSize, 10.ToString());
+			configuration.SetProperty(Cfg.Environment.UseConnectionOnSystemTransactionPrepare, true.ToString());
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from System.Object").ExecuteUpdate();
+
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void ShouldWorkWithOuterSystemTransaction()
+		{
+			if (!Sfi.ConnectionProvider.Driver.SupportsSystemTransactions)
+				Assert.Ignore("System.Transactions support is required by this test");
+
+			const int count = 3;
+			var entities = new List<TestEntity>(count);
+			for (var i = 0; i < count; i++)
+			{
+				entities.Add(new TestEntity { Id = Guid.NewGuid() });
+			}
+
+			using (var transaction = new TransactionScope())
+			{
+				using (var session = Sfi.OpenStatelessSession())
+				{
+					foreach (var entity in entities)
+					{
+						session.Insert(entity);
+					}
+				}
+
+				transaction.Complete();
+			}
+
+			using (var session = OpenSession())
+			{
+				var results = session.QueryOver<TestEntity>().List<TestEntity>();
+
+				Assert.That(results.Count, Is.EqualTo(count));
+			}
+		}
+
+		[Test]
+		public void ShouldWorkWithInnerSystemTransaction()
+		{
+			if (!Sfi.ConnectionProvider.Driver.SupportsSystemTransactions)
+				Assert.Ignore("System.Transactions support is required by this test");
+
+			const int count = 3;
+			var entities = new List<TestEntity>(count);
+			for (var i = 0; i < count; i++)
+			{
+				entities.Add(new TestEntity { Id = Guid.NewGuid() });
+			}
+
+			using (var session = Sfi.OpenStatelessSession())
+			{
+				using (var transaction = new TransactionScope())
+				{
+					foreach (var entity in entities)
+					{
+						session.Insert(entity);
+					}
+					session.FlushBatcher();
```

Comment:
`StatelessSession` has `ManagedFlush` method which is not exposed through interface. Maybe we should expose it instead?

Reply:
I choosed `FlushBatcher` rather than just `Flush` because in principle the stateless session should never need a `Flush`, since it is stateless. So I want to stress what this flush is for: flushing the batcher that can be enabled with the stateless session, and which may retain some operations not yet sent to the database.

Reply:
Just wanted to clarify that I'm not talking about the general `Flush` method, but about `ManagedFlush`.

Reply:
In the case of the stateless session, that is the same.

But anyway, what does it mean, `ManagedFlush`? I do not make any sense of it. `FlushBatcher` is explicit about its purpose, while that is not the case of `ManagedFlush` in my opinion.

https://github.com/nhibernate/nhibernate-core/blob/2bd3c1680e105dc496a4c97f6d8630ca362feee8/src/NHibernate/Impl/StatelessSessionImpl.cs#L385-L396